### DISCUSSION TITLE
Idiom for left join?

### DISCUSSION THREAD

**Discussion by mwotton** [TimefoldAI/timefold-solver#748] [General]
I find myself needing to test that all Foos have at least n associated Bars. In SQL I'd do a left join followed by a group with a count projection, but I'm not sure what the right way to do this in timefold is (as the obvious join and group fails in the case that a Foo has zero Bars). I can sort of see how to cobble it together with one stream that collects it into a list, and concat-ing it together with another stream that simply provides all Foos with an empty list of Bars, then grouping on Foo.id, but it seems a bit inelegant: is there a better construction?

**Comment by mwotton** [comment]
Is the answer just to split the constraint into "not enough"and "zero" variants?

**Comment by triceo** [comment]
Hello @mwotton! Before the existence of `concat`, the only way to do this was - as you suggest - to have two different constraints. However, `concat` was introduced precisely so that you can do this. The constraint could look something like this:

    forEach(Foo.class)
    .join(Bar.class, 
        equal(Function.identity(), Bar::getFoo))
    .concat(forEach(Foo.class)
         .ifNotExists(Bar.class,
             equal(Function.identity(), Bar::getFoo)))
    .groupBy((foo, bar) -> foo,
        ConstraintCollectors.conditionally(
            (foo, bar) -> bar != null,
            ConstraintCollectors.countBi()))
    .filter((foo, barCount) -> ...)
    ...
